File: crates/taflow-core/src/indicators/force_index.rs
use crate::error::TaResult;
use crate::indicators::ExponentialMovingAverage;
use crate::stream::StreamingIndicator;

/// EMA-smoothed close-to-close price force weighted by volume.
#[derive(Debug, Clone)]
pub struct ForceIndex {
    previous_close: Option<f64>,
    average: ExponentialMovingAverage,
    count: usize,
    value: Option<f64>,
}

impl ForceIndex {
    /// Create a Force Index with the requested EMA smoothing period.
    pub fn new(period: usize) -> TaResult<Self> {
        Ok(Self {
            previous_close: None,
            average: ExponentialMovingAverage::new(period)?,
            count: 0,
            value: None,
        })
    }

    /// Append one close/volume pair and return the smoothed force when ready.
    pub fn append(&mut self, close: f64, volume: f64) -> Option<f64> {
        self.count += 1;
        let Some(previous) = self.previous_close.replace(close) else {
            return None;
        };
        self.value = self.average.append((close - previous) * volume);
        self.value
    }

    /// Return the latest smoothed force, or `None` during warm-up.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Return the number of processed close/volume pairs.
    pub fn len(&self) -> usize {
        self.count
    }

    /// Return whether no pairs have been processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }

    /// Clear previous-close, EMA, count, and latest-value state.
    pub fn reset(&mut self) {
        self.previous_close = None;
        self.average.reset();
        self.count = 0;
        self.value = None;
    }
}

File: python/taflow/indicators/force_index.py
"""Native-backed Force Index adapter."""
from typing import Any

import numpy as np

from .._native import ForceIndexOperator as _Native
from .._series import as_float64_series


class ForceIndex:
    """Compute EMA-smoothed signed price force weighted by volume.

    ``close`` and ``volume`` are required equal-length chronological series
    and may both be empty for a fresh stream. Rust computes
    ``EMA((close_t - close_(t-1)) * volume_t, period)``. The first ``period``
    bars are warm-up NaN because one bar seeds the prior close and ``period``
    force observations seed the EMA. Wickra ``ForceIndex`` is the oracle.
    ``compute`` returns one aligned float array, ``value`` is the latest scalar
    or ``None`` during warm-up, and lifecycle mutators return ``self``.
    """

    def __init__(self, period: int = 13) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native(period)

    def append(self, close: float, volume: float) -> "ForceIndex":
        """Append one close/volume pair and return this adapter."""
        self._state.append(float(close), float(volume))
        return self

    def extend(self, close: Any, volume: Any) -> "ForceIndex":
        """Append equal-length close and volume histories."""
        arrays = as_float64_series(close), as_float64_series(volume)
        if len(arrays[0]) != len(arrays[1]):
            raise ValueError("close and volume must have equal lengths")
        self._state.extend(*arrays)
        return self

    def compute(self) -> np.ndarray:
        """Return the aligned Force Index history."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest force value, or ``None`` during warm-up."""
        return self._state.value

    def reset(self) -> "ForceIndex":
        """Restore fresh native state and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of processed pairs."""
        return len(self._state)
